site-src/geps/gep-1058.md

```diff
@@ -0,0 +1,962 @@
+The second use case is allowing the owner of a route to delegate responsibility
+for portions of the route configuration.
+
+This proposal is specifically focused on httpRoutes but creates a design
+pattern that can be applied to other route types in the future. For this reason,
+this GEP primarily uses the generic term “Route” instead of using httpRoute.
+
+## Use Cases
+
+There are two primary use cases that this proposal is intended to address. There
+are also a number of secondary use cases that are a combination of the two
+primary ones. The two primary use cases are described below.
+
+The diagram below is an example scenario that is referred to
+throughput this document and the example configurations are based on it.
+
+**Example Scenario 1:**  
+![Example Scenario 1](images/1058-example-scenario.png "Example Scenario 1")
+
+### 1. Partial Administrative Delegation of Routing
+
+There are scenarios where there is a need to divide responsibility for route
+configuration across multiple users and/or groups.
+
+Using Scenario 1 as an example, there are four groups of administrative users,
+one for each name space. The `prod-gw` namespace is managed by the NetOps team
+and the other three admin groups are the Application Owners, each with their own
+namespace.
+
+The NetOps team has a single Listener that handles all HTTP-based traffic for
+the host www.example.com but they want to manage route configuration only enough
+to ensure traffic is sent to the correct namespace. The NetOps team wants to
+delegate the rest of the route configuration to the application owners.
+
+It should be noted that both the NetOps team and the Application teams may need
+to configure more features of httpRoute than just matching rules. Examples of
+this are header manipulation and path rewriting.
+
+### 2. Compose routes from reusable config fragments
+
+Another use case is where there is a group of route config settings that are
+common across multiple Routes of the same type (e.g. httpRoute). Users want to
+be able to have the Routes include that group of config settings by reference
+instead of having to embed those settings directly in each Route or RouteRule.
+
+The configurations for Scenario 1 (shown lower on this page), demonstrate two
+ways this technique can be used. In the first case, traffic needs to be routed
+to "Core Services" (`core-svces-ns`) from two different starting points. Using
+route inclusion, the routing configuration for the Core Services are put in a
+child route and then that child route is included in two sections of a parent
+route where they are needed.
+
+Another way route inclusion is used in Scenario 1 is by configuring a
+HTTPRouteFilter, that inserts a header, in the parent route. This results in the
+filter being added to each RouteRule in the child route `store-route`. The child
+route is both shorter and easier to read than it would have been without this
```

Comment:
> The important part of having the resources be in separate namespaces is not that they're shorter, it's that they can be owned by different people, and those people can coordinate without having to coordinate edits of a single resource.

I agree. There are different personas that delegation is used for.

Reply:
@youngnick I 100% agree with that use case and I think is a good motivation for this GEP. I think that is "Partial Administrative Delegation of Routing" though - you are delegating to different roles. This section ("Compose routes from reusable config fragments") seems to be suggesting it should be used as just a way to DRY config within the same ownership which I don't agree is a good idea.

The design may make this usage _possible_ (I think we would have to go out of our way to stop it, if we wanted, by not allowing same-namespace delegation, which seems odd), but I don't think we should list it as a use case or encourage it.

Totally possible I am missing more context on why this is useful though.

Reply:
This is very important - 'different personas' also means different roles and privileges. The Gateway admin has control over the host DNS and certificates, and can decide which namespaces to trust and for what hostname and path prefix, and can set policies for the host.

Maintaining security and separation of roles should be a clear requirement for any delegation proposal. My concern with the complexity of this proposal is in large part because it makes the security properties hard to understand - which will lead to CVEs.